### PR TITLE
Fix nth week holidays

### LINKED ISSUE #41274: Have a happy Thanksgiving in the United States! (Holiday system ignores week parameter it seems)
Issue reported from  Round ID: 96414 (/tg/Station Sybil [ENGLISH] [US-EAST] [100% LAG FREE])
Reporting client version: 512

### LINKED ISSUE #51652: Holidays with nth day of a certain type this month don't work
Father's day is meant to be the third Sunday in June, but the game thinks it's today because it's the third week in June, but only the second Sunday since the month started on a Monday.

### PR DESCRIPTION
**Pull request by jacklecroy**
## Changelog
:cl: That REALLY Good Soda Flavor
fix: Holidays appearing on the nth day of the month like Thanksgiving will work now.
tweak: Columbus day is now called Indigenous Peoples' Day.
/:cl:

Fixes #51652
Fixes #41274

### CODE REVIEW

File: code/modules/unit_tests/holidays.dm [hunk 1]
```diff
@@ -0,0 +1,35 @@
+// test Jewish holiday
+/datum/unit_test/hanukkah_2123/Run()
+	var/datum/holiday/hebrew/hanukkah/hanukkah = new
+	TEST_ASSERT_EQUAL(hanukkah.shouldCelebrate(13, DECEMBER, 2123, 2, MONDAY), TRUE, "December 13, 2123 was not Hannukkah.")
```
Comment (Jared-Fogle) [inline code/modules/unit_tests/holidays.dm:4]: For this and everything else, use `TEST_ASSERT` when checking if something is TRUE or FALSE.

For example--`TEST_ASSERT(hannukah.shouldCelebrate(...), "This wasn't Hannukah")` for checking TRUE, and `TEST_ASSERT(!holiday.shouldCelebrate(...), "Holiday was celebrated when it shouldn't have been")` for checking FALSE.

File: code/modules/unit_tests/holidays.dm [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+// test Islamic holiday
+/datum/unit_test/ramadan_4208/Run()
+	var/datum/holiday/islamic/ramadan/ramadan = new
+	TEST_ASSERT_EQUAL(ramadan.shouldCelebrate(1, MARCH, 4208, 1, TUESDAY), TRUE, "March 1, 4208 was not the start of Ramadan.")
+
+// there were problems on this day
+/datum/unit_test/nov_1_2018/Run()
+	for(var/H in subtypesof(/datum/holiday))
+		var/datum/holiday/holiday = new H()
+		TEST_ASSERT_EQUAL(holiday.shouldCelebrate(1, NOVEMBER, 2018, 1, TUESDAY), FALSE, "A holiday tried to celebrate on November 1, 2018.")
```
Comment (Jared-Fogle) [inline code/modules/unit_tests/holidays.dm:15]: This is a very strange, and non forwards compatible test! What's the deal with November 1st, 2018, and why should no *future* holidays be allowed to be celebrated on that day?

**Review by Mothblocks — CHANGES_REQUESTED** [review]

File: code/modules/unit_tests/holidays.dm (hunk 2 above)
Comment (jacklecroy) [reply code/modules/unit_tests/holidays.dm:15]: True

File: code/modules/holiday/nth_week.dm [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+	drone_hat = /obj/item/clothing/head/that //This is the closest we can get to a pilgrim's hat
+
+/datum/holiday/nth_week/thanksgiving/canada
+	name = "Thanksgiving in Canada"
+	begin_week = 2
+	begin_month = OCTOBER
+	begin_weekday = MONDAY
+
+/datum/holiday/nth_week/indigenous
+	// not Columbus day anymore get rekt, Columbus
```
Comment (Jared-Fogle) [inline code/modules/holiday/nth_week.dm:54]: based

**Comment by jacklecroy** [comment]
Well something is messed up

File: code/modules/holiday/nth_week.dm (hunk 3 above)
Comment (Jared-Fogle) [reply code/modules/holiday/nth_week.dm:54]: but put it in the changelog

**Comment by jacklecroy** [comment]
Oh yeah foreign calendars use realtime I gotta change that

**Comment by Jared-Fogle** [comment]
You can test on your own system before committing by uncommenting `//#define UNIT_TESTS` in _compile_options.dm, then running tgstation.dmb with Dream Daemon.

**Comment by jacklecroy** [comment]
Date should be better for foreign calendar because of floating point precision with realtime. Also it can be unit tested now without having to calculate the realtime which is difficult to calculate.

**Comment by jacklecroy** [comment]
First day of month calculation was really wrong

**Comment by jacklecroy** [comment]
Ready to merge.

File: code/__HELPERS/time.dm [hunk 4]
```diff
@@ -66,28 +66,15 @@ GLOBAL_VAR_INIT(rollovercheck_last_timeofday, 0)
...
-		if(3)
-			day = 2
-		if(4)
-			day = 3
-		if(5)
-			day = 4
-		if(6)
-			day = 5
-	return day
+/proc/first_day_of_month(year, month)
```
Comment (Jared-Fogle) [inline code/__HELPERS/time.dm:69]: Document. I know you didn't make it, but you remade it.

File: code/modules/holiday/foreign_calendar.dm [hunk 5]
```diff
@@ -1,28 +1,46 @@
...
-		jd = realtime_to_jd()
+		jd = gregorian_to_jd(yy, mm, dd)
 	set_date(jd)
 
 /datum/foreign_calendar/proc/set_date()
 	return
 
-/datum/foreign_calendar/proc/realtime_to_jd()
-	return round(world.realtime / 864000) + BYOND_EPOCH
+/datum/foreign_calendar/proc/gregorian_to_jd(year, month, day)
```
Comment (Jared-Fogle) [inline code/modules/holiday/foreign_calendar.dm:25]: Document.

File: code/modules/holiday/foreign_calendar.dm [hunk 6]
```diff
@@ -1,28 +1,46 @@
 #define BYOND_EPOCH 2451544.5
 #define HEBREW_EPOCH 347995.5
 #define ISLAMIC_EPOCH 1948439.5
+#define GREGORIAN_EPOCH 1721424.5
```
Comment (Jared-Fogle) [inline code/modules/holiday/foreign_calendar.dm:4]: #undef all these at the end.

**Review by Mothblocks — CHANGES_REQUESTED** [review]
Last bit--need documentation on the new stuff you added or completely redid. I noticed you did some of them, but it's required for all new procs and such.

https://github.com/tgstation/tgstation/blob/master/.github/AUTODOC_GUIDE.md

File: code/__HELPERS/time.dm (hunk 4 above)
Comment (jacklecroy) [reply code/__HELPERS/time.dm:69]: already is documented

**Review by Mothblocks — APPROVED** [review]

**Comment by Jared-Fogle** [comment]
Thanks for the tests!

**Pull request merged**